bluepysnap/edges.py

```diff
@@ -480,7 +482,37 @@ def _get(self, selection, properties=None):
 
         return result
 
-    def ids(self, group=None, limit=None, sample=None):
+    def _edge_ids_by_filter(self, queries, raise_missing_prop):
+        """Return edge IDs if their properties match the `queries` dict.
+
+        `props` values could be:
+            pairs (range match for floating dtype fields)
+            scalar or iterables (exact or "one of" match for other fields)
+
+        You can use the special operators '$or' and '$and' also to combine different queries
+        together.
+
+        Examples:
+            >>> self._edge_ids_by_filter({ Edge.POST_SECTION_ID: (0, 1),
+            >>>                            Edge.AXONAL_DELAY: (.5, 2.) })
+            >>> self._edge_ids_by_filter({'$or': [{ Edge.PRE_X_CENTER: [2, 3]},
+            >>>                              { Edge.POST_SECTION_POS: (0, 1),
+            >>>                              Edge.SYN_WEIGHT: (0.,1.4) }]})
+
+        """
+        properties = query.get_properties(queries)
+        unknown_props = properties - self.property_names
+        if raise_missing_prop and unknown_props:
+            raise BluepySnapError(f"Unknown edge properties: {unknown_props}")
+        res = []
+        ids = self.ids(None)
+        chunk_size = int(1e8)
```

Comment:
We can keep it like this for the moment but do you think we can keep it close to ~10Gb of memory usage in the future ?

Reply:
Do you mean to detect the amount of available RAM by cleaver usage?

Reply:
yes I was writing it here : 
https://github.com/BlueBrain/snap/pull/129#discussion_r586322548

What do you think ?

Reply:
I don t have a super strong opinion. I just hate when my pc freeze when it was not clearly expected.

Reply:
Please confirm
- detect available RAM memory
- detect size of edges file
- raise an exception if RAM is too low
- otherwise calculate chunk size, and start processing

Reply:
I would say yes with these functions : get_available_ram(), get_dataframe_memory() in utils.py maybe.

But what do you think about it ? Because this is doable but do you think this is "elegant" or even useful ? or should we just don t care about that ?

Reply:
In general it is useful. However it is ahead of time in my opinion.

Reply:
RAM checks

Reply:
Ok fine with me

Reply:
I'm still going to implement them here :)

Reply:
ok, I've checked with this code

```python
        from psutil import virtual_memory, swap_memory

        properties = properties - unknown_props
        ids = self.ids(None)
        available_memory = swap_memory().free + virtual_memory().available
        max_required_memory = len(properties) * 8 * len(ids)  # assume 8 bytes per edge value
        if available_memory < max_required_memory:
            raise BluepySnapError('Not enough memory for edges querying')
        print(f'available_memory {available_memory}, max_required_memory {max_required_memory}')
        chunk_size = int(1e9)
        for chunk in np.array_split(ids, 1 + len(ids) // chunk_size):
            data = self.get(chunk, properties)
            res.extend(chunk[query.resolve_ids(data, self.name, queries)])
        return np.array(res, dtype=IDS_DTYPE)
```

It does not work because `libsonata` fails before we even reach this. So if there is not enough memory then `libsonata` will fail at
```
    @cached_property
    def _population(self):
        return self._edge_storage.storage.open_population(self.name)
```
Not much point in this check because if there is not enough memory then we don't reach it.

Reply:
ok really good to know.

Reply:
Thanks for the check. 
I must say I am not sure what we can do to avoid filling memory at the moment. So let s merge it like this and we will see. 
Maybe reducing the chunk size to 1e7 ?

Reply:
ok but I want to emphasize again that if there is not enough memory => we fail at `libsonata` usage before reaching this code. Basically when we in `def _edge_ids_by_filter`, we already have enough memory.

Reply:
ahhh ok. Sorry that all good then !

Reply:
:+1:

Reply:
Thank you so much. It was very helpful and useful.